src/NHibernate/Type/GenericSetType.cs

```diff
@@ -76,5 +78,15 @@ public override object Instantiate(int anticipatedSize)
 		{
 			return new HashSet<T>();
 		}
+
+		protected override bool AreCollectionElementsEqual(IEnumerable original, IEnumerable target)
```

Comment:
I have thought about it, but I have seen many entity classes with set property loosely typed `ICollection` and initialized with something else than a set (a `List<>`by example) and then added to the session: it works. Changing the code above could cause it to no more work.